mongo/description/server_selector.go

```diff
@@ -185,47 +185,48 @@ func selectForReplicaSet(rp *readpref.ReadPref, isOutputAggregate bool, t Topolo
 		return nil, err
 	}
 
+	// If underlying operation is an aggregate with an output stage, only apply read preference
+	// if all candidates are 5.0+. Otherwise, operate under primary read preference.
```

Comment:
I need to think about this a bit more, but leaving a comment here to start.

I think logic in `Operation.createReadPref` for applying a read preference to an operation may need to updated too to match the specification.

From this:

```
if desc.Server.Kind == description.Standalone || (isOpQuery && desc.Server.Kind != description.Mongos) ||
    op.Type == Write || (op.IsOutputAggregate && desc.Server.WireVersion.Max < 13) {
    // Don't send read preference for:
    // 1. all standalones
    // 2. non-mongos when using OP_QUERY
    // 3. all writes
    // 4. when operation is an aggregate with an output stage, and selected server's wire
    //    version is < 13
    return nil, nil
}
```

To something like:

```
if desc.Server.Kind == description.Standalone || (isOpQuery && desc.Server.Kind != description.Mongos) ||
    op.Type == Write || (op.IsOutputAggregate && (/* TODO */)) {
    // Don't send read preference for:
    // 1. all standalones
    // 2. non-mongos when using OP_QUERY
    // 3. all writes
    // 4. when operation is an aggregate with an output stage, and the read preference was overridden to primary
    return nil, nil
}
```

I am not sure how feasible that is to change or test. I am not sure if that has any visible impact.

The only scenario I think this could surface is with a mixed version replica set like:

- Primary with maxWireVersion=12
- Secondary with maxWireVersion=13
- User uses readPref=Secondary
- `Operation.createReadPref` applies `$readPreference: secondary`.

If that has no impact on correctness (i.e. the primary does not error at the passed `$readPreference`) that may not be a blocking issue.

Reply:
Offline discussion made me realize an incorrect `$readPreference` will affect command monitoring events as well. I am of the same stance. I prefer fixing this now. But if it is complex, I think this will not be a significant enough issue to block the 1.8.0 release.

Reply:
I've filed GODRIVER-2231 to correctly report `$readPreference` when the supplied read preference was overwritten in `operation.createReadPref`. 

I don't _think_ it's trivial to determine whether read preference was overwritten [here](https://github.com/mongodb/mongo-go-driver/pull/804/files#diff-23f4872f37e48a31948f158719eaaeda2cf0dc5f24931c95f54c2ed6daa7d0c7R190-R196) in `operation.go`. Server selection does not return any information besides a list of candidate servers and an error. We could store some global variable, but global mutable state is of course undesirable. We may need a more clever solution here. Happy to brainstorm offline, but I agree that this shouldn't block the 1.8.0 release. Sending a "secondary" read preference to a Primary is not an error. I think the only risk would be a user seeing the incorrect "secondary" read preference in command monitoring.

Reply:
SGTM. Thank you for filing GODRIVER-2231.